### ISSUE TITLE
use webjars

### ISSUE THREAD

**Issue by nazoking** [gitbucket/gitbucket#1430]
Would you like to use webjars?
http://www.webjars.org/

Webjars has libraries of npm and bower registered, you can easily use it from servlet by just registering it in sbt. Many registered javascript and css are optimized. (refs #965 Compress JavaScript and CSS resources)

Also, by using webjars, it is guaranteed that the original is unchanged, so it helps to upgrade the library version.

**Comment by aadrian** [comment]
using npm and webpack directly (from gradle or sbt) offer all the advantages and many more IMHO.

**Comment by McFoggy** [comment]
Is this issue to get rid of `gitbucket/src/main/webapp/assets/vendors/` and to allow somehow an automatic dependency managment towards JS/CSS libs during buildtime?

**Comment by aadrian** [comment]
> Is this issue to get rid of gitbucket/src/main/webapp/assets/vendors/ and to allow somehow an automatic dependency managment towards JS/CSS libs during buildtime?

If this is the case, it can be achieved wihout WebJars very easily: just have a simple ```package.json``` with the needed dependencies, and let NPN take care of the download und update. Than with WebPack https://webpack.js.org/ the usage can  be optimized (much better than it can be with WebJars) .

**Comment by nazoking** [comment]
> using npm and webpack directly (from gradle or sbt) offer all the advantages and many more IMHO.

I agree. But webjars are very simple. There is no need to set anything. Npm and webpack require new software and configuration files for development. It is somewhat complicated to use from sbt, exhausting developers. If we aim for Single Page Application, webpack looks best. However, webjars works well and it is easy to migrate if we use the current usage of javascript.

**Comment by nazoking** [comment]
> Is this issue to get rid of gitbucket/src/main/webapp/assets/vendors/ and to allow somehow an automatic dependency managment towards JS/CSS libs during buildtime?

yes. i would rid gitbucket/src/main/webapp/assets/vendors/ .

**Comment by nazoking** [comment]
webjars is not software. it is library for scala or java, servlet. We need no sbt-plugins.

The work is all just the following.

  1. add script name to `dependencies` in `build.sbt`
       * like `"org.webjars.nmp" % "bootstrap" % "3.3.6"`
  2. grep and rewrite path in script or link tag.
       * ex. `@helpers.assets/vendors/bootstrap-3.3.6/css/bootstrap.min.css` to `/webjars/bootstrap/3.3.6/css/bootstrap.min.css` in [main.scala.html](https://github.com/gitbucket/gitbucket/blob/72049c5bdf0796b83df6584e3179ecae8708580a/src/main/twirl/gitbucket/core/main.scala.html#L12)
       * Perhaps we should add a new helper function to indicate the path of webjars like `@helpers.assets`
  3. delete `gitbucket/src/main/webapp/assets/vendors/bootstrap-3.3.6`
  4. it is all

**Comment by ritschwumm** [comment]
@aadrian i really like having projects build out-of-the-box without being forced to install and maintain more than one build-/package-management tool

**Comment by takezoe** [comment]
+1 for webjars. GitBucket is still using Vanilla.js and jQuery. It's not riding on the modern front-end development, so introducing other solutions like npm or webpack is overkill. I think webjars will fit for GitBucket and it will simply decrease our costs to manage JavaScript libraries.

**Comment by aadrian** [comment]
@ritschwumm you don't need to, as the build tools can take care of everything. In both cases (webjars or npm) the only maintenance work is to change in some file the "version number" to a newer version .

> ... so introducing other solutions like npm or webpack is overkill.

@takezoe npm and webpack are **build** time artifacts, not runtime, so the produced output is much more optimized - something that WebJars simply can't achieve (especially useful for the mobile use)

> I think webjars will fit for GitBucket and it will simply decrease our costs to manage JavaScript libraries.

WebJars are indeed a nice improvement so that those resources are fetched at build time, and don't make the GIT repo unnecessarily heavy.

> GitBucket is still using Vanilla.js and jQuery. It's not riding on the modern front-end development, ..

Well, to me it seems that GitBucket is using quite a few JS and CSS resources :) : 
https://github.com/gitbucket/gitbucket/tree/master/src/main/webapp/assets/vendors
(and with more plug-ins even more JS and CSS resources will be required ) .

Since only parts of those many resources are really used, but they're loaded, I mentioned ```webpack``` since it would open the possibility that we later optimize the runtime footprint to the really used resource code parts.

**Comment by ritschwumm** [comment]
@aadrian so when i have sbt installed i don't need to install npm and webpack on my machine to use them? i doubt it.

**Comment by aadrian** [comment]
> i don't need to additionally install npm and webpack on my machine to use them? i doubt it.

@ritschwumm You can (the 1st time build will be faster), but you don't have to :) .

Gradle can do everything required behind the scenes (download node, download npms, call tasks, all the same way it's doing with regular JARs) with https://github.com/srs/gradle-node-plugin/blob/master/docs/installing.md  or many other plug-ins. 
SBT should be no different here, since there are even very specialized plug-ins like https://github.com/stejskal/sbt-webpack .

The only scenario where this doesn't seem to work smoothly yet is when working behind corporate dynamic proxies: Gradle (Java) doesn't seem to pass correctly the required settings to the node/npm/bower/grunt processes .

**Comment by t-tsutsumi** [comment]
You can see some progress in #1595.

**Comment by yaroot** [comment]
Can we at least switch to LFS if no progress is made in in switching to webjars or node-related tools?
